### PR TITLE
Condense enqueue_ems_action

### PR DESCRIPTION
**Pull request by kbrock**
Blocked by / built on top of:
- [x] https://github.com/ManageIQ/manageiq-api/pull/1103 (first 2 commits)

### Overview

We have a pattern for actions that request the ems to perform an operation.
So I converted all `queue_object_action` calls over to `enqueue_ems_action`

Suggest reviewing by commit (unsure if commits will show in proper order but they are contained)

### Before
- `enqueue_action` mostly works for ems operations
- some ems operations get a user and others do not

### After
- fixed queue messaging when auto deriving a description.
- `enqueue_action` is generic and `enqueue_ems_action` is ems specific.
- user is now sent to all ems operations.
- introduced `:supports` parameter to easily enforce `supports?`.
- everything has been converted across except for subcollections and generic objects.
 

### Review

Only the first 2 commits do anything, and are minor changes at that.
The subsequent commits just leverage `enqueue_ems_action` (and delete code)

---

### Code effect

```diff
 class CloudTemplatesController < BaseController
   def check_compliance_resource(type, id, _data = nil)
-    api_action(type, id) do |klass|
-      container_image = resource_search(id, type, klass)
-      api_log_info("Checking compliance of #{container_image_ident(container_image)}")
-      request_compliance_check(container_image)
-   end
+   enqueue_ems_action(type, id, "Check Compliance for", :method_name => "check_compliance", :supports => true)
  end
-
-  private
-
-  def cloud_template_ident(cloud_template)
-    "CloudTemplate id:#{cloud_template.id} name:'#{cloud_template.name}'"
-  end
-
-  def request_compliance_check(cloud_template)
-    desc = "#{cloud_template_ident(cloud_template)} check compliance requested"
-    raise BadRequestError, "Checking Compliance for #{cloud_template_ident(cloud_template)}: #{model.unsupported_reason(:check_compliance)}" unless model.supports?(:check_compliance)
-
-    task_id = queue_object_action(cloud_template, desc, :method_name => "check_compliance")
-    action_result(true, desc, :task_id => task_id)
-  rescue StandardError => err
-    action_result(false, err.to_s)
-  end
 end
```

Of course, not all are as compact as this, but this is a common and best case scenario.

### CODE REVIEW

File: app/controllers/api/base_controller/action.rb
Comment (Fryguy) [inline app/controllers/api/base_controller/action.rb:38]: offhand this looks like it could be DRYed up with the code in the other rescue block.

File: app/controllers/api/base_controller/action.rb
Comment (Fryguy) [inline app/controllers/api/base_controller/action.rb:78]:
```suggestion
        hash = action_phrase.kind_of?(Hash) ? action_phrase : options
        hash.reverse_merge!(:role => "ems_operations", :user => true)
```

File: app/controllers/api/vms_controller.rb
Comment (Fryguy) [inline app/controllers/api/vms_controller.rb:119]: Saying we're losing something in a comment is confusing, because a reader doesn't know what we are losing it relative to (relative to the previous code or relative to something earlier in the current code).  Instead, I think a TODO about what are our intentions for the code is easier to understand.

```suggestion
      # TODO: Find a way to get `new_name` into the response text
```

Comment (kbrock) [reply app/controllers/api/vms_controller.rb]: the consolidation is going to loos stuff

File: app/controllers/api/vms_controller.rb
Comment (kbrock) [reply app/controllers/api/vms_controller.rb:120]: There are a few instances of these.

I had debated this.

For a request with bad parameters , this will still look up the vm.
But we still want the block to log the result, add the resource link, and depending if it is a single resource, raise the exception or just use a failed action response hash.;

I can dive in and see if there is a dry way to handle this.

**Comment by kbrock** [comment]
@Fryguy ok, reduced the duplicate code.

I want to continue going with reducing the `api_action` work next (probably another 🍪  worth)

So let me know your thoughts on this one and I can plan accordingly

**Comment by kbrock** [comment]
codeclimate: 45 closed, one opened

File: app/controllers/api/vms_controller.rb
Comment (kbrock) [reply app/controllers/api/vms_controller.rb]: yikes. this method should not have been refactored

Comment (kbrock) [reply app/controllers/api/vms_controller.rb]: Also, I solved this in the PR I just extracted. At least for the case where we are not queuing

**Comment by kbrock** [comment]
PR is good, blocked by dependent PR that was pulled out out

Don't want to review that code too many times

File: app/controllers/api/base_controller/action.rb
Comment (Fryguy) [inline app/controllers/api/base_controller/action.rb:91]: Since we require `:user => true` via the interface and technically that value can be overridden...we should check the actual value.  (i.e. `{:user => false}.key?(:user) == true`

```suggestion
        if options[:user]
```

**Comment by Fryguy** [comment]
Overall LGTM - only one comment.  I'd like @agrare to also review before merge.

**Review by agrare — APPROVED** [review]
:+1: from me looks awesome

File: app/controllers/api/base_controller/action.rb
Comment (kbrock) [reply app/controllers/api/base_controller/action.rb]: done

**Comment by kbrock** [comment]
push: fixed edge case in deriving `action_phrase` from `enqueue_ems_action`

**Pull request merged**
